src/Controller/AppController.js

```diff
@@ -1,58 +1,27 @@
 // Utilities
 import * as utils from "./utils";
-import { countries } from "./constants";
+import { openaqAPI_REQ, wikiAPI_REQ } from "./constants";
 // External dependencies
 import { trackPromise } from "react-promise-tracker";
 
 // Fetch OpenAQ API data on form submit.
-export function getAPIData(
-  event,
-  apiStateSetter,
-  formInputState,
-  formInputSetter,
-  formValidationSetter
-) {
-  apiStateSetter([]); // clear data on submit;
-  event.preventDefault();
-  if (
-    utils.isCountrySupported(countries, formInputState, formValidationSetter)
-  ) {
-    const code = utils.getCountryCode(formInputState);
-    const openaqAPI_KEY = `https://api.openaq.org/v1/measurements?country=${code}&parameter=pm25&order_by[]=value&order_by[]=city&sort[]=desc&has_geo=true&value_from=1&date_from=${utils.getDateDaysAgo(
-      7
-    )}`;
-    trackPromise(
-      fetch(openaqAPI_KEY)
-        .then(res => res.json())
-        .then(json => utils.filterDataByCity(json.results)) // get the top 10 polluted cities in given country
-        .then(
-          pollutedCities =>
-            fetchCityDescription(
-              pollutedCities,
-              apiStateSetter,
-              formInputState,
-              formInputSetter
-            ) // fetch descriptions from WikiAPI.)
-        )
-    );
-  } else apiStateSetter([]);
+export function getAPIData(countryCode) {
+  return trackPromise(
+    fetch(openaqAPI_REQ(countryCode))
+      .then(response => response.json())
+      .then(data => utils.filterDataByCity(data.results)) // get the top 10 polluted cities in given country
+      .then(pollutedCities => fetchCityDescription(pollutedCities))
+  );
 }
 
 // Fetch city description from WikiAPI and add to Top10 object properties.
-function fetchCityDescription(
-  cityArray,
-  apiStateSetter,
-  formInputState,
-  formInputSetter
-) {
-  cityArray.forEach(item => {
-    const wikiAPI_KEY = `https://en.wikipedia.org/w/api.php?origin=*&action=query&prop=extracts&format=json&exintro=true&titles=${item.city}`; // Grab the first page section extract.
-    fetch(wikiAPI_KEY)
-      .then(res => res.json())
-      .then(data => {
-        item.cityDescription = utils.extractDescriptionFromWiki(data); // Add extract text to Top10 object.
-        apiStateSetter(cityArray);
-        utils.forceRerender(formInputState, formInputSetter); // force a rerender to make sure city descriptions are updated in the <p> tags.
-      });
+function fetchCityDescription(data) {
+  const citiesWithTrivia = JSON.parse(JSON.stringify(data)); // deep clone to avoid mutation
+  citiesWithTrivia.forEach(measurement => {
```

Comment:
```suggestion
  data.forEach(measurement => {
```
forEach doesn't mutate your array so there is no need of cloning it